UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AuthTokenProvider.cs

```diff
@@ -1,30 +1,86 @@
+        private readonly ILogger<AuthTokenProvider> logger;
+        private static readonly Object _lock = new Object();
+        private readonly IDistributedCache _cache;
 
-        public AuthTokenProvider(IOptions<EssManagedIdentityConfiguration> essManagedIdentityConfiguration)
+        public AuthTokenProvider(IOptions<EssManagedIdentityConfiguration> essManagedIdentityConfiguration, IDistributedCache _cache, ILogger<AuthTokenProvider> logger)
         {
             this.essManagedIdentityConfiguration = essManagedIdentityConfiguration;
+            this._cache = _cache;
+            this.logger = logger;
         }
 
+
         public async Task<string> GetManagedIdentityAuthAsync(string resource)
+        {
+            var accessToken = GetFromCache(resource);
+
+            if (accessToken != null && accessToken.AccessToken != null && accessToken.ExpiresIn > DateTime.UtcNow)
+            {
+                return accessToken.AccessToken;
+            }
+
+            var newAccessToken = await GetNewAuthToken(resource);
+            AddToCache(resource, newAccessToken);
+
+            return newAccessToken.AccessToken;
+        }
+
+        private async Task<AccessTokenItem> GetNewAuthToken(string resource)
         {
             var tokenCredential = new DefaultAzureCredential(new DefaultAzureCredentialOptions { ManagedIdentityClientId = essManagedIdentityConfiguration.Value.ClientId });
             var accessToken = await tokenCredential.GetTokenAsync(
                 new TokenRequestContext(scopes: new string[] { resource + "/.default" }) { }
             );
 
-            return accessToken.Token;
+            return new AccessTokenItem
+            {
+                ExpiresIn = accessToken.ExpiresOn.UtcDateTime,
+                AccessToken = accessToken.Token
+            };
+        }
+
+        private void AddToCache(string key, AccessTokenItem accessTokenItem)
+        {
+            var tokenExpiryMinutes = accessTokenItem.ExpiresIn.Subtract(DateTime.UtcNow).TotalMinutes;
+            var deductTokenExpiryMinutes = essManagedIdentityConfiguration.Value.DeductTokenExpiryMinutes < tokenExpiryMinutes ? essManagedIdentityConfiguration.Value.DeductTokenExpiryMinutes : 1;
+            var options = new DistributedCacheEntryOptions().SetSlidingExpiration(TimeSpan.FromMinutes(tokenExpiryMinutes - deductTokenExpiryMinutes));
+            options.SetAbsoluteExpiration(accessTokenItem.ExpiresIn);
+
+            lock (_lock)
+            {
+                _cache.SetString(key, JsonConvert.SerializeObject(accessTokenItem), options);
+                logger.LogInformation(EventIds.CachingExternalEndPointToken.ToEventId(), "Caching new token for external end point resource {resource} and expires in {ExpiresIn} with sliding expiration duration {options}.", key, Convert.ToString(accessTokenItem.ExpiresIn), JsonConvert.SerializeObject(options));
+            }
+        }
+
+        private AccessTokenItem GetFromCache(string key)
```

Comment:
Kindly change name of method from **GetFromCache** To **GetAuthTokenFromCache** for more clarity.

Reply:
Yes did the changes accordingly